### PR TITLE
Fix issue looking up routes that don't exist

### LINKED ISSUE #59: Registering multiple routes with the same prefix unintentionally creates new routes
Simplified case from fastify/fastify#786

```js
var router = require('find-my-way')()

router.on('GET', '/b/bulk', function b_bulk () {})

console.log(router.find('GET', '/bulk')) // null

router.on('GET', '/b/', function b () {})

console.log(router.find('GET', '/bulk')) // { handler: [Function: b_bulk], params: {}, store: null }
```

Somehow the `POST /bulk` route exists after the `POST /b/` route is registered.

`router.prettyPrint()` doesn't show that it exists though:

```
└── /
    └── b/ (GET)
        └── bulk (GET)
```

Looks like it has something to do with the fact that `b/` and `bulk` both start with the letter `b`. If you change `bulk` to something that doesn't start with the letter `b`, it works as expected.

### LINKED ISSUE #60: Found a wrong route
The following code finds a wrong route
```js
const noop = () => {}
const findMyWay = FindMyWay()

findMyWay.on('GET', '/bb/', noop)
findMyWay.on('GET', '/bb/bulk', noop)

const f = findMyWay.find('GET', '/bulk')

// f === noop
```

Expected: 
`f` should be null

Issue born here: https://github.com/fastify/fastify/pull/787

### PR DESCRIPTION
**Pull request by nwoltman**
How this fix works:

When matching a path to a prefix in a `node`, this part of the code finds the longest common prefix. Before it would only look for the next part of the path if the path matched the entire prefix, otherwise it would continue matching the same path. This was wrong because then this would happen:

```
path = 'bulk'
prefix = 'b/'
commonPrefix = 'b'
-- does not update path --
finds child with prefix = 'bulk'
goes back to the top and matches 'bulk'
return handler for '/b/bulk'
```

With this fix it does this:

```
path = 'bulk'
prefix = 'b/'
commonPrefix = 'b'
update path = 'ulk'
no children match 'ulk'
return null
```

<details>
<summary>benchmarks</summary>

**before**
```
lookup static route x 26,167,641 ops/sec ±1.42% (89 runs sampled)
lookup dynamic route x 1,568,707 ops/sec ±0.56% (93 runs sampled)
lookup dynamic multi-parametric route x 872,150 ops/sec ±1.18% (90 runs sampled)
lookup dynamic multi-parametric route with regex x 800,702 ops/sec ±1.98% (83 runs sampled)
lookup long static route x 2,584,658 ops/sec ±1.07% (92 runs sampled)
find static route x 32,397,375 ops/sec ±1.69% (89 runs sampled)
find dynamic route x 1,942,803 ops/sec ±1.57% (88 runs sampled)
find dynamic multi-parametric route x 1,066,697 ops/sec ±1.29% (88 runs sampled)
find dynamic multi-parametric route with regex x 926,740 ops/sec ±1.48% (90 runs sampled)
find long static route x 4,905,622 ops/sec ±1.39% (89 runs sampled)
```

**after**
```
lookup static route x 27,301,612 ops/sec ±1.26% (88 runs sampled)
lookup dynamic route x 1,576,759 ops/sec ±0.55% (94 runs sampled)
lookup dynamic multi-parametric route x 890,252 ops/sec ±0.56% (94 runs sampled)
lookup dynamic multi-parametric route with regex x 845,854 ops/sec ±0.43% (93 runs sampled)
lookup long static route x 2,425,642 ops/sec ±0.64% (93 runs sampled)
find static route x 30,674,123 ops/sec ±0.81% (88 runs sampled)
find dynamic route x 1,875,931 ops/sec ±0.76% (93 runs sampled)
find dynamic multi-parametric route x 1,037,566 ops/sec ±0.63% (92 runs sampled)
find dynamic multi-parametric route with regex x 955,495 ops/sec ±0.90% (94 runs sampled)
find long static route x 4,807,380 ops/sec ±0.67% (94 runs sampled)
```
</details>
<br>

Alternative to #61.

Fixes #59
Fixes #60

### CODE REVIEW

**Comment by nwoltman** [comment]
@mcollina I updated the PR description.

File: index.js [hunk 1]
```diff
@@ -300,7 +300,7 @@ Router.prototype.find = function find (method, path) {
     i = pathLen < prefixLen ? pathLen : prefixLen
     while (len < i && path[len] === prefix[len]) len++
 
-    if (len === prefixLen) {
+    if (len > 0) {
```
Comment (allevo) [inline index.js]: `if (len === prefixLen && len > 0)`?

Comment (nwoltman) [reply index.js]: Nope. The problem is the `len === prefixLen` part.

Comment (nwoltman) [reply index.js]: I think maybe `len === prefixLen` is correct and something else needs to change.

**Comment by nwoltman** [comment]
@delvedor I added that test. I think I don't understand the code well enough to fix it.

**Comment by delvedor** [comment]
Add `if ((isStatic === true && len > 0) || len === prefixLen) {` fixes everything, but there is a small regression in the performances.
Apparently if we are working with a static route the length of the common prefix just need to be higher than 0, while if we are working with a dynamic route it **must** be the same of the prefix length.

**Comment by delvedor** [comment]
I've fixed the test.
I'll open another pr with some optimization that will remove the little regression that we are having here.

File: test/issue-59.test.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+})
+
+test('static / 2', t => {
+  t.plan(1)
+  const findMyWay = FindMyWay()
+
+  findMyWay.on('GET', '/bb/ff/', noop)
+  findMyWay.on('GET', '/bb/ff/bulk', noop)
+
+  t.equal(findMyWay.find('GET', '/bulk'), null)
```
Comment (allevo) [inline test/issue-59.test.js]: Could you add /ff/bulk?

Comment (delvedor) [reply test/issue-59.test.js]: Done

Comment (allevo) [reply test/issue-59.test.js]: I mean in the assertion

Comment (delvedor) [reply test/issue-59.test.js]: Done :D

**Comment by delvedor** [comment]
Landed in `v1.10.2`, thanks!

**Comment by nwoltman** [comment]
So it turns out that my idea with `len > 0` wasn't good enough. With `v1.10.2` this case fails:

```js
var router = require('find-my-way')()

router.on('GET', '/b/', function b () {})
router.on('GET', '/b/bulk', function b_bulk () {})
router.on('GET', '/b/ulk', function b_ulk () {})

console.log(router.find('GET', '/bulk')) // { handler: [Function: b_ulk], ... }
```

**Comment by delvedor** [comment]
@nwoltman that's weird!
Since is a very edge case I'll consider this bug as fixed, let's open another issue for the test case you have just found :)

**Comment by allevo** [comment]
ops... too late
